### PR TITLE
fix(rum-react): render the correct component when using ApmRoute with Switch

### LINKED ISSUE #340: make Apmroute component work with functional components
While debugging a react issue, eventually ran in to an issue where our `ApmRoute` component does not work with functional components which has effects. A simple example to reproduce. 

```js
import React, { useEffect, useState } from 'react'

export default function funcComponent() {
  const [count, setCount] = useState(0)

  useEffect(() => {
    async function fetchDummy() {
      await fetch('./dummy')
      setCount(2)
    }
    fetchDummy()
  })

  return <div>{count}</div>
}


<ApmRoute path="/func" component={funcComponent} />
```

### PR DESCRIPTION
**Pull request by vigneshshanmugam**
+ fixes elastic/apm-agent-rum-js#343 
+ restructured the e2e tests a bit to add more test cases. 
+ add functional component called via `useEffect` pattern

### CODE REVIEW

**Comment by vigneshshanmugam** [comment]
Test fails because of buid_docs which is unrelated to the PR.

File: packages/rum-react/src/get-with-transaction.js [hunk 1]
```diff
@@ -23,20 +23,23 @@
...
 function getWithTransaction(apm) {
   return function withTransaction(name, type) {
-    return function(WrappedComponent) {
-      if (!WrappedComponent) {
-        return WrappedComponent
+    return function(Component) {
+      if (!Component) {
+        const loggingService = apm.serviceFactory.getService('LoggingService')
+        loggingService.warn(
+          `ApmRoute is not instrumenting the route since component property is not provided`
```
Comment (jahtalab) [inline packages/rum-react/src/get-with-transaction.js]: This probably shouldn't mention "route"

File: packages/rum-react/test/e2e/react/app.jsx [hunk 2]
```diff
@@ -76,50 +72,49 @@ class App extends React.Component {
...
-          />
-          <ApmRoute path="/home" component={MainComponent} />
-          <Route path="/about" render={() => <div>about</div>} />
-          <ApmRoute path="/topics" component={MainComponent} />
-          <ApmRoute path="/topic/:id" component={TopicComponent} />
-          <Route
-            path="/manual/"
-            component={() => {
-              return (
+          <Switch>
```
[the remark this answers is not in the document]
Comment (vigneshshanmugam) [reply packages/rum-react/test/e2e/react/app.jsx]: Sure let me create a separate E2E for this.

File: packages/rum-react/src/get-with-transaction.js (hunk 1 above)
Comment (vigneshshanmugam) [reply packages/rum-react/src/get-with-transaction.js]: will change the naming. good catch.

File: packages/rum-react/src/get-apm-route.js [hunk 3]
```diff
@@ -29,24 +29,44 @@ import { getWithTransaction } from './get-with-transaction'
...
+
   return class ApmRoute extends React.Component {
     constructor(props) {
       super(props)
-      const { path, component: Component } = this.props
-      this.ApmComponent = withTransaction(path, 'route-change')(Component)
+      const { path, component } = this.props
+      this.state = {
+        path,
+        apmComponent: withTransaction(path, 'route-change')(component)
```
[the remark this answers is not in the document]
Comment (vigneshshanmugam) [reply packages/rum-react/src/get-apm-route.js]: We can keep either do this or other way it to keep the keep it in the instance and update. 

```
constructor {
  this.state = { path }
  this.apmComponent = withTransaction(path, 'route-change')(component)
}

static getDerivedStateFromProps(nextProps, prevState) {
      const { component, path } = nextProps

      if (path !== prevState.path) {
         this.apmComponent = withTransaction(path, 'route-change')(component)
         return { path }
      }

      return null
    }

```

**Pull request merged**